packages/@ember/-internals/runtime/lib/system/array_proxy.ts

```diff
@@ -306,8 +305,14 @@ class ArrayProxy<T, C extends EmberArray<T> = EmberArray<T>>
       // @ts-expect-error This check is still good for ensuring correctness
       assert("Can't set ArrayProxy's content to itself", arrangedContent !== this);
       assert(
-        `ArrayProxy expects an Array or ArrayProxy, but you passed ${typeof arrangedContent}`,
-        isArray(arrangedContent) || (arrangedContent as any).isDestroyed
+        `ArrayProxy expects a native Array, EmberArray, or ArrayProxy, but you passed ${typeof arrangedContent}`,
+        (function (arr: unknown): arr is EmberArray<unknown> {
+          return Array.isArray(arr) || EmberArray.detect(arr);
+        })(arrangedContent)
```

Comment:
Ha! I independently came up with the type-narrowing-in-`assert`-IIFE around the same time.